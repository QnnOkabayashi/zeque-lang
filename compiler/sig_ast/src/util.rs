use std::hash::Hash;
use std::iter::{Enumerate, Rev};
use std::ops::{Deref, Index, IndexMut};
use std::{cmp, fmt, hash, marker::PhantomData, slice::Iter};

pub type StringInterner = string_interner::StringInterner<string_interner::backend::BufferBackend>;

#[derive(Copy, Clone, Debug)]
pub struct Span<T>(pub T, pub Range);

impl<T> Span<T> {
    pub fn new(start: usize, end: usize, inner: T) -> Self {
        Span(
            inner,
            Range {
                start: start as u32,
                end: end as u32,
            },
        )
    }

    pub fn range(&self) -> Range {
        self.1
    }

    pub fn into_inner(self) -> T {
        self.0
    }
}

/// PartialEq ignores range
impl<T: PartialEq> PartialEq for Span<T> {
    fn eq(&self, other: &Self) -> bool {
        self.0 == other.0
    }
}

impl<T: Eq> Eq for Span<T> {}

/// Hash ignores range
impl<T: Hash> Hash for Span<T> {
    fn hash<H: hash::Hasher>(&self, state: &mut H) {
        self.0.hash(state);
    }
}

#[derive(Copy, Clone, Debug)]
pub struct Range {
    pub start: u32,
    pub end: u32,
}

impl Range {
    pub fn to(self, other: Self) -> Self {
        debug_assert!(self.end <= other.start, "unexpected use of Range::to");
        Range {
            start: self.start,
            end: other.end,
        }
    }
}

// invariant: values and ranges are the same length
#[derive(Clone, Debug)]
pub struct RangeTable<T> {
    values: Vec<T>,
    ranges: Vec<Range>,
}

impl<T> RangeTable<T> {
    pub fn new() -> Self {
        RangeTable::default()
    }

    pub fn ranges(&self) -> &[Range] {
        &self.ranges
    }

    pub fn span(&self, index: Ix<T>) -> Span<&T> {
        Span(&self.values[index], self.ranges[index.index])
    }

    pub fn push(&mut self, Span(value, range): Span<T>) -> Ix<T> {
        self.ranges.push(range);
        Ix::push(&mut self.values, value)
    }
}

impl<T> Default for RangeTable<T> {
    fn default() -> Self {
        RangeTable {
            values: Vec::new(),
            ranges: Vec::new(),
        }
    }
}

impl<T> Deref for RangeTable<T> {
    type Target = [T];

    fn deref(&self) -> &Self::Target {
        &self.values
    }
}

pub struct Scope<'scope, T> {
    original_len: usize,
    stack: &'scope mut Vec<T>,
}

impl<T> Drop for Scope<'_, T> {
    fn drop(&mut self) {
        self.stack.truncate(self.original_len);
    }
}

impl<'scope, T> Scope<'scope, T> {
    pub fn new(env: &'scope mut Vec<T>) -> Self {
        Scope {
            original_len: env.len(),
            stack: env,
        }
    }

    pub fn enter_scope(&mut self) -> Scope<'_, T> {
        Scope {
            original_len: self.stack.len(),
            stack: self.stack,
        }
    }

    pub fn push(&mut self, value: T) {
        self.stack.push(value);
    }

    /// Includes the enumerate part in case people want to use the index...
    /// This is a _stack_. We search _backwards_. So if you want the position, you need to
    /// make sure the index from enumerate reflects that, which it does if .rev() comes
    /// _after_ .enumerate().
    pub fn iter(&self) -> Rev<Enumerate<Iter<'_, T>>> {
        self.stack.iter().enumerate().rev()
    }

    pub fn get(&self, index: usize) -> &T {
        &self.stack[index]
    }

    pub fn get_mut(&mut self, index: usize) -> &mut T {
        &mut self.stack[index]
    }

    pub fn in_scope(&self) -> &[T] {
        &self.stack[self.original_len..]
    }
}

impl<T> Extend<T> for Scope<'_, T> {
    fn extend<I: IntoIterator<Item = T>>(&mut self, iter: I) {
        self.stack.extend(iter)
    }
}

/// An index type.
pub struct Ix<I: ?Sized> {
    pub index: usize,
    _marker: PhantomData<I>,
}

impl<I> Ix<I> {
    pub fn new(index: usize) -> Self {
        Ix {
            index,
            _marker: PhantomData,
        }
    }

    pub fn map<U>(self) -> Ix<U> {
        Ix::new(self.index)
    }

    /// Push a value and get an index to it.
    pub fn push<T>(vec: &mut Vec<T>, value: T) -> Self
    where
        I: Indexes<T>,
    {
        let this = Ix::new(vec.len());
        vec.push(value);
        this
    }

    /// Iterate over the indices of a slice.
    pub fn iter(slice: &[I]) -> impl Iterator<Item = Ix<I>> {
        (0..slice.len()).map(Ix::new)
    }

    pub fn enumerate(iter: &[I]) -> impl Iterator<Item = (Self, &I)> {
        iter.iter()
            .enumerate()
            .map(|(index, value)| (Ix::new(index), value))
    }
}

impl<T> Copy for Ix<T> {}

impl<T> Clone for Ix<T> {
    fn clone(&self) -> Self {
        *self
    }
}

impl<T> fmt::Debug for Ix<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Ix<{}>({})", std::any::type_name::<T>(), self.index)
    }
}

impl<T> fmt::Display for Ix<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

impl<T> cmp::PartialEq for Ix<T> {
    fn eq(&self, other: &Self) -> bool {
        self.index == other.index
    }
}

impl<T> cmp::Eq for Ix<T> {}

impl<T> hash::Hash for Ix<T> {
    fn hash<H: hash::Hasher>(&self, state: &mut H) {
        self.index.hash(state);
    }
}

/// If Foo implements Indexes<Bar>, then an Ix<Foo> may be used
/// to index into a &[Bar].
/// All types are allowed to index into slices containing themselves.
pub trait Indexes<T> {}

impl<T> Indexes<T> for T {}

// Can index into a slice of options,
// e.g. Ix<Function> can index into &[Option<Function>]
impl<T> Indexes<Option<T>> for T {}

impl<T, I: Indexes<T>> Index<Ix<I>> for [T] {
    type Output = T;

    fn index(&self, index: Ix<I>) -> &Self::Output {
        &self[index.index]
    }
}

impl<T, I: Indexes<T>> IndexMut<Ix<I>> for [T] {
    fn index_mut(&mut self, index: Ix<I>) -> &mut Self::Output {
        &mut self[index.index]
    }
}

impl<T, I: Indexes<T>> Index<Ix<I>> for Vec<T> {
    type Output = T;

    fn index(&self, index: Ix<I>) -> &Self::Output {
        self.as_slice().index(index)
    }
}

impl<T, I: Indexes<T>> IndexMut<Ix<I>> for Vec<T> {
    fn index_mut(&mut self, index: Ix<I>) -> &mut Self::Output {
        self.as_mut_slice().index_mut(index)
    }
}
