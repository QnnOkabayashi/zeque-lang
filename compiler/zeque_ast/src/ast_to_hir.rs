//! The AST -> HIR transformation consists of:
//!   * name resolution
//!   * using an array and indices for the tree structure

use crate::util::Scope;
use crate::{ast, hir};
use index_vec::IndexVec;
use smol_str::SmolStr;
use thiserror::Error;

#[derive(Clone, Debug, Error)]
pub enum Error {
    #[error("name not found: {0}")]
    NameNotFound(String),
}

pub fn entry(file: &ast::File) -> Result<hir::File, Error> {
    let mut env = vec![];
    let mut scope = Scope::new(&mut env);

    let mut lowerer = LoweringContext::new(hir::Level::new(0));
    let struct_idx = lowerer.lower_decls(&file.decls, &mut scope)?;

    Ok(hir::File {
        struct_idx,
        ctx: lowerer.ctx,
    })
}

struct LoweringContext {
    ctx: hir::Ctx,
    level: hir::Level,
}

impl LoweringContext {
    fn new(level: hir::Level) -> Self {
        LoweringContext {
            ctx: hir::Ctx::default(),
            level,
        }
    }

    fn lower_expr_raw(
        &mut self,
        expr: &ast::Expr,
        scope: &mut Scope<'_, (SmolStr, hir::Local)>,
    ) -> Result<hir::Expr, Error> {
        match expr {
            ast::Expr::Int(int) => Ok(hir::Expr::Int(*int)),
            ast::Expr::Bool(boolean) => Ok(hir::Expr::Bool(*boolean)),
            ast::Expr::BinOp { op, lhs, rhs } => {
                let lhs = self.lower_expr(lhs, scope)?;
                let rhs = self.lower_expr(rhs, scope)?;

                Ok(hir::Expr::BinOp { op: *op, lhs, rhs })
            }
            ast::Expr::IfThenElse { cond, then, else_ } => {
                let cond = self.lower_expr(cond, scope)?;
                let then = self.lower_expr(then, scope)?;
                let else_ = self.lower_expr(else_, scope)?;

                Ok(hir::Expr::IfThenElse { cond, then, else_ })
            }
            ast::Expr::Name(name) => {
                let name = self.lower_name(name.clone(), scope)?;

                Ok(hir::Expr::Name(name))
            }
            ast::Expr::Call { callee, args } => {
                let callee = match callee {
                    ast::Callee::Expr(callee) => {
                        let callee_index = self.lower_expr(callee, scope)?;
                        hir::Callee::Expr(callee_index)
                    }
                    ast::Callee::Builtin(builtin_name) => {
                        let builtin = builtin_name.parse().unwrap_or_else(|e| match e {});

                        hir::Callee::Builtin(builtin)
                    }
                };

                let args = args
                    .iter()
                    .map(|argument| self.lower_expr(argument, scope))
                    .collect::<Result<_, _>>()?;

                Ok(hir::Expr::Call { callee, args })
            }
            ast::Expr::Block(block) => {
                let block_index = self.lower_block(block, scope)?;

                Ok(hir::Expr::Block(block_index))
            }
            ast::Expr::Comptime(expr) => {
                let expr_index = self.lower_expr(expr, scope)?;

                Ok(hir::Expr::Comptime(expr_index))
            }
            ast::Expr::Struct(struct_) => {
                let struct_index = self.lower_struct(struct_, scope)?;

                Ok(hir::Expr::Struct(struct_index))
            }
            ast::Expr::Constructor { ty, fields } => {
                let ty = ty
                    .as_deref()
                    .map(|ty| self.lower_expr(ty, scope))
                    .transpose()?;

                let fields = fields
                    .iter()
                    .map(|field| {
                        let value = self.lower_expr(&field.expr, scope)?;

                        Ok(hir::ConstructorField {
                            name: field.name.clone(),
                            value,
                        })
                    })
                    .collect::<Result<_, Error>>()?;

                Ok(hir::Expr::Constructor { ty, fields })
            }
            ast::Expr::FieldAccess { expr, field_name } => {
                let expr = self.lower_expr(expr, scope)?;

                Ok(hir::Expr::Field {
                    expr,
                    field_name: field_name.clone(),
                })
            }
        }
    }

    fn lower_expr(
        &mut self,
        expr: &ast::Expr,
        scope: &mut Scope<'_, (SmolStr, hir::Local)>,
    ) -> Result<hir::ExprIdx, Error> {
        let expr = self.lower_expr_raw(expr, scope)?;
        let expr_idx = self.ctx.exprs.push(expr);
        Ok(expr_idx)
    }

    fn lower_param(
        &mut self,
        param: &ast::Param,
        scope: &mut Scope<'_, (SmolStr, hir::Local)>,
    ) -> Result<hir::ParamIdx, Error> {
        let ty = self.lower_expr(&param.ty, scope)?;
        let param_idx = self.ctx.params.push(hir::Param {
            is_comptime: param.is_comptime.is_some(),
            name: param.name.clone(),
            ty,
        });
        scope.push((
            param.name.clone(),
            hir::Local {
                level: self.level,
                kind: hir::LocalKind::Param(param_idx),
            },
        ));
        Ok(param_idx)
    }

    fn lower_block(
        &mut self,
        block: &ast::Block,
        scope: &mut Scope<'_, (SmolStr, hir::Local)>,
    ) -> Result<hir::Block, Error> {
        let mut scope = scope.enter_scope();
        let mut stmts = Vec::with_capacity(block.stmts.len());
        for stmt in &block.stmts {
            stmts.push(self.lower_stmt(stmt, &mut scope)?);
        }

        let returns = self.lower_expr(&block.returns, &mut scope)?;
        Ok(hir::Block { stmts, returns })
    }

    fn lower_stmt(
        &mut self,
        stmt: &ast::Stmt,
        scope: &mut Scope<'_, (SmolStr, hir::Local)>,
    ) -> Result<hir::Stmt, Error> {
        match stmt {
            ast::Stmt::Let(let_) => {
                let expr = self.lower_expr(&let_.expr, scope)?;
                let ty = let_
                    .ty
                    .as_ref()
                    .map(|ty| self.lower_expr(&ty, scope))
                    .transpose()?;

                let let_idx = self.ctx.lets.push(hir::Let {
                    name: let_.name.clone(),
                    ty,
                    expr,
                });

                scope.push((
                    let_.name.clone(),
                    hir::Local {
                        level: self.level,
                        kind: hir::LocalKind::Let(let_idx),
                    },
                ));

                Ok(hir::Stmt::Let(let_idx))
            }
        }
    }

    fn lower_name(
        &mut self,
        name: SmolStr,
        scope: &mut Scope<'_, (SmolStr, hir::Local)>,
    ) -> Result<hir::Name, Error> {
        for (scoped_name, local) in scope.iter() {
            if scoped_name.as_str() == name.as_str() {
                return Ok(hir::Name::Local(*local));
            }
        }

        if let Ok(builtin_type) = name.parse() {
            return Ok(hir::Name::BuiltinType(builtin_type));
        }

        Err(Error::NameNotFound(name.to_string()))
    }

    fn lower_struct(
        &mut self,
        struct_: &ast::Struct,
        scope: &mut Scope<'_, (SmolStr, hir::Local)>,
    ) -> Result<hir::StructIdx, Error> {
        let mut scope = scope.enter_scope();
        self.lower_decls(&struct_.decls, &mut scope)
    }

    fn lower_decls(
        &mut self,
        decls: &[ast::Decl],
        scope: &mut Scope<'_, (SmolStr, hir::Local)>,
    ) -> Result<hir::StructIdx, Error> {
        // insert a placeholder value
        let struct_idx = self.ctx.structs.push(hir::Struct::default());

        let mut num_fns = 0;
        let mut num_fields = 0;
        // Put the names of all the fns in scope (and count so we can preallocate too!)
        for decl in decls {
            match decl {
                ast::Decl::Fn(fn_decl) => {
                    scope.push((
                        fn_decl.name.clone(),
                        hir::Local {
                            level: self.level,
                            kind: hir::LocalKind::Fn(struct_idx, hir::FnIdx::new(num_fns)),
                        },
                    ));
                    num_fns += 1;
                }
                ast::Decl::Field(_) => {
                    num_fields += 1;
                }
            }
        }

        let mut struct_ = hir::Struct {
            fns: IndexVec::with_capacity(num_fns),
            fields: Vec::with_capacity(num_fields),
        };

        for decl in decls {
            match decl {
                ast::Decl::Fn(fn_decl) => {
                    struct_.fns.push(self.lower_fn_decl(fn_decl, scope)?);
                }
                ast::Decl::Field(field_decl) => {
                    struct_
                        .fields
                        .push(self.lower_field_decl(field_decl, scope)?);
                }
            }
        }

        self.ctx.structs[struct_idx] = struct_;

        Ok(struct_idx)
    }

    fn lower_fn_decl(
        &mut self,
        fn_decl: &ast::FnDecl,
        scope: &mut Scope<'_, (SmolStr, hir::Local)>,
    ) -> Result<hir::FnDecl, Error> {
        let mut lowerer = LoweringContext::new(hir::Level::new(self.level.index() + 1));

        let mut params = Vec::with_capacity(fn_decl.params.len());
        for param in &fn_decl.params {
            params.push(lowerer.lower_param(param, scope)?);
        }

        let return_ty = lowerer.lower_expr(&fn_decl.return_ty, scope)?;
        let body = lowerer.lower_block(&fn_decl.body, scope)?;

        Ok(hir::FnDecl {
            is_pub: fn_decl.is_public.is_some(),
            name: fn_decl.name.clone(),
            params,
            return_ty,
            body,
            ctx: lowerer.ctx,
        })
    }

    fn lower_field_decl(
        &mut self,
        field_decl: &ast::FieldDecl,
        scope: &mut Scope<'_, (SmolStr, hir::Local)>,
    ) -> Result<hir::FieldDecl, Error> {
        let ty = self.lower_expr(&field_decl.ty, scope)?;
        Ok(hir::FieldDecl {
            name: field_decl.name.clone(),
            ty,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    macro_rules! snapshot {
        ($($tokens:tt)*) => {{
            insta::assert_debug_snapshot!(parse(stringify!($($tokens)*)));
        }};
    }

    fn parse(source: &str) -> hir::File {
        let ast = crate::parse::file(source).expect("a valid ast");
        entry(&ast).expect("a valid hir")
    }

    #[test]
    fn one_function() {
        snapshot! {
            fn x() i32 {
                1
            }
        }
    }

    #[test]
    fn stmt() {
        snapshot! {
            fn x() i32 {
                let y = 4;
                y
            }
        }
    }

    #[test]
    fn field_decl() {
        snapshot! {
            a: i32,
        }
    }

    #[test]
    fn struct_def() {
        snapshot! {
            a: struct {
                b: i32,
            },
        }
    }

    #[test]
    fn nested_ctxs() {
        snapshot! {
            fn foo() type {
                struct {
                    fn bar() type {
                        i32
                    }
                }
            }
        }
    }

    #[test]
    fn reference_outer_ctxs() {
        snapshot! {
            fn foo(comptime T: type) type {
                struct {
                    fn bar() type {
                        T
                    }
                }
            }
        }
    }

    #[test]
    fn function_call() {
        snapshot! {
            fn main() i32 {
                foo()
            }

            fn foo() i32 {
                1
            }
        }
    }

    #[test]
    fn if_then_else() {
        snapshot! {
            fn main() i32 {
                if true { 1 } else { 0 }
            }
        }
    }

    #[test]
    fn precedence() {
        snapshot! {
            fn main() bool {
                1 + 2 * 3 == 4
            }
        }
    }
}
