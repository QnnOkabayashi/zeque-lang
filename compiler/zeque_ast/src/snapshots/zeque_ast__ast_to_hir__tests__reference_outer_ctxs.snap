---
source: compiler/zeque_ast/src/ast_to_hir.rs
expression: "parse(stringify!(fn foo(comptime T: type) type\n        { struct { fn bar() type { T } } }))"
---
Hir {
    structs: [
        Struct {
            fns: [
                FnDecl {
                    is_pub: false,
                    name: "foo",
                    params: [
                        0,
                    ],
                    return_ty: 1,
                    body: Block {
                        stmts: [],
                        returns: 2,
                    },
                    ctx: Ctx {
                        params: [
                            Param {
                                is_comptime: true,
                                name: "T",
                                ty: 0,
                            },
                        ],
                        exprs: [
                            BuiltinType(
                                Type,
                            ),
                            BuiltinType(
                                Type,
                            ),
                            Struct(
                                1,
                            ),
                        ],
                        lets: [],
                        parent_captures: [],
                    },
                },
            ],
            fields: [],
        },
        Struct {
            fns: [
                FnDecl {
                    is_pub: false,
                    name: "bar",
                    params: [],
                    return_ty: 0,
                    body: Block {
                        stmts: [],
                        returns: 1,
                    },
                    ctx: Ctx {
                        params: [],
                        exprs: [
                            BuiltinType(
                                Type,
                            ),
                            Name(
                                ParentRef(
                                    0,
                                ),
                            ),
                        ],
                        lets: [],
                        parent_captures: [
                            Local(
                                Param(
                                    0,
                                ),
                            ),
                        ],
                    },
                },
            ],
            fields: [],
        },
    ],
    files: [
        File {
            struct_idx: 0,
            ctx: Ctx {
                params: [],
                exprs: [],
                lets: [],
                parent_captures: [],
            },
        },
    ],
    main: 0,
}
